tests/testthat/test-model-compile-user_header.R

```diff
@@ -0,0 +1,104 @@
+
+file_that_exists <- 'placeholder_exists'
+file_that_doesnt_exist <- 'placeholder_doesnt_exist'
+file.create(file_that_exists)
+on.exit(if(file.exists(file_that_exists)) file.remove(file_that_exists), add=TRUE, after=FALSE)
+
+make_local_orig <- cmdstan_make_local()
+cmdstan_make_local(cpp_options = list("PRECOMPILED_HEADERS"="false"))
+on.exit(cmdstan_make_local(cpp_options = make_local_orig, append = FALSE), add = TRUE, after = FALSE)
+
+test_that("cmdstan_model works with user_header with mock", {
+  skip_if(os_is_macos())
+  tmpfile <- tempfile(fileext = ".hpp")
+  hpp <-
+  "
+  #include <stan/math.hpp>
+  #include <boost/math/tools/promotion.hpp>
+  #include <ostream>
+
+  namespace bernoulli_external_model_namespace
+  {
+      template <typename T0__,
+            stan::require_all_t<stan::is_stan_scalar<T0__>>* = nullptr>
+      inline typename boost::math::tools::promote_args<T0__>::type make_odds(const T0__ &
+                                                                                 theta,
+                                                                             std::ostream *pstream__)
+      {
+          return theta / (1 - theta);
+      }
+  }"
+  cat(hpp, file = tmpfile, sep = "\n")
+
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_mock_compile(
+    expect_warning(
+      expect_no_warning({
+        mod <- cmdstan_model(
+          stan_file = testing_stan_file("bernoulli_external"),
+          exe_file = file_that_exists,
+          user_header = tmpfile
+        )
+      }, message = 'Recompiling is recommended'), # this warning should not occur because recompile happens automatically
+      'Retrieving exe_file info failed' # this warning should occur
+    )
+  ))
+
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_mock_compile({
+    mod_2 <- cmdstan_model(
+      stan_file = testing_stan_file("bernoulli_external"),
+      exe_file = file_that_doesnt_exist,
+      cpp_options=list(USER_HEADER=tmpfile),
+      stanc_options = list("allow-undefined")
+    )
+  }))
+
+  # Check recompilation upon changing header
+  file.create(file_that_exists)
+  with_mocked_cli(compile_ret = list(status = 0), info_ret = list(), code = expect_no_mock_compile({
```

Comment:
We haven't been running any linters on the code, although we could start doing that. Either way, I agree about keeping the lines short. There are likely some other places in the code where that needs to be fixed, although I've tried to keep them under 80 in the code I've written.

Reply:
yeah, since it doesn't seem like a linter was run in the past, I didn't run one on my code either, but that makes it surely inconsistent. I think there is a way to run linter just on the diff so at least we don't make the problem worse. I can look into it.

Reply:
> I think there is a way to run linter just on the diff so at least we don't make the problem worse. I can look into it.

If that's simple to do then that sounds good, otherwise don't worry about it. We can (and probably should) start running a linter on the whole package, maybe after merging this PR.

Reply:
TBH - I wouldn't advise that. It basically creates a huge diff so your subsequent blames become less meaningful. I think there is a way to exclude certain commits from blame...but....

Reply:
Hmm yeah that’s a good point

Reply:
We did this years ago in the stan math library. We essentially have a huge file that tells the compiler to ignore any auto changes.

For cmdstanr we could set up a hook so that the linter runs before any future commits. So we would put the one global change in the `.git-blame-ignore-revs` then have a git hook that runs precommit to run the linter

https://github.com/stan-dev/math/blob/develop/.git-blame-ignore-revs

But again, seperate issue from the PR

@katrinabrock if you can try to be mindful of not having too long of lines that would be great, but no need to go through counting column

Reply:
I went ahead and linted all my changes.